pkg/openai/openai.go

```diff
@@ -3,18 +3,23 @@ package openai
 import (
 	"context"
 	"fmt"
-	"math"
 	"net/url"
+	"unicode/utf8"
 
-	"github.com/pandodao/tokenizer-go"
+	"github.com/pkoukk/tiktoken-go"
 	"github.com/sashabaranov/go-openai"
 )
 
 type Client struct {
-	OpenAIClient *openai.Client
+	tiktokenEncoding *tiktoken.Tiktoken
+	OpenAIClient     *openai.Client
 }
 
-func NewClient(apiSecret string, apiHost string) *Client {
+func NewClient(apiSecret string, apiHost string) (*Client, error) {
+	tokenizer, err := tiktoken.EncodingForModel(openai.GPT3Dot5Turbo)
```

Comment:
Is `*tiktoken.Tiktoken` concurrent-safe?

Reply:
hmmm, tried to read the code. The `EncodingForModel(...)` calls `GetEncoding(..)`, then it calls `getEncoding(...)`, however, `getEncoding(...)` had a map write, so I think it is definitely not safe to call `EncodingForModel(...)` in Goroutines, but the rest of the member methods that `*tiktoken.Tiktoken` had seem to be ok to call in Goroutines.

Reply:
I think so, encoding/decoding does not change *tiktoken.Tiktoken itself, but I'm not sure.